### PR TITLE
[Flax] Add FlaxMBart

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?
This PR adds flax implementation of MBart.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten  @patil-suraj

### CODE REVIEW

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 1]
```diff
@@ -0,0 +1,1732 @@
...
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs
+
+
+class FlaxMBartEncoderLayerCollection(nn.Module):
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:450]: Can we add copied from here?

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 2]
```diff
@@ -0,0 +1,1732 @@
...
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (self_attn_weights, cross_attn_weights)
+
+        return outputs
+
+
+class FlaxMBartDecoderLayerCollection(nn.Module):
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:592]: Can we add copied from here?

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 3]
```diff
@@ -0,0 +1,1732 @@
...
+    config: MBartConfig
+    dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+
+    def setup(self):
+        self.layers = [
+            FlaxMBartDecoderLayer(self.config, name=str(i), dtype=self.dtype)
+            for i in range(self.config.decoder_layers)
+        ]
+        self.layerdrop = self.config.encoder_layerdrop
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:601]:
```suggestion
        self.layerdrop = self.config.decoder_layerdrop
```

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 4]
```diff
@@ -0,0 +1,1732 @@
...
+
+
+class FlaxMBartEncoder(nn.Module):
+    config: MBartConfig
+    dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+    embed_tokens: Optional[nn.Embed] = None
+
+    def setup(self):
+        self.dropout_layer = nn.Dropout(rate=self.config.dropout)
+        self.layerdrop = self.config.encoder_layerdrop
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:699]: think we can remove this. We don't need layerdrop here.

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 5]
```diff
@@ -0,0 +1,1732 @@
...
+
+
+class FlaxMBartDecoder(nn.Module):
+    config: MBartConfig
+    dtype: jnp.dtype = jnp.float32  # the dtype of the computation
+    embed_tokens: Optional[nn.Embed] = None
+
+    def setup(self):
+        self.dropout_layer = nn.Dropout(rate=self.config.dropout)
+        self.layerdrop = self.config.decoder_layerdrop
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:773]: same comment as above.

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 6]
```diff
@@ -0,0 +1,1732 @@
...
+
+        return FlaxBaseModelOutputWithPastAndCrossAttentions(
+            last_hidden_state=outputs.last_hidden_state,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            cross_attentions=outputs.cross_attentions,
+        )
+
+
+class FlaxMBartModule(nn.Module):
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:850]: Think we can add copied from here as well.

File: tests/test_modeling_flax_mbart.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+        )
+
+        outputs = model.decode(decoder_input_ids, encoder_outputs, decoder_attention_mask=decoder_attention_mask)
+
+        diff = np.max(np.abs((outputs_cache_next[0][:, -1, :5] - outputs[0][:, -1, :5])))
+        self.parent.assertTrue(diff < 1e-3, msg=f"Max diff is {diff}")
+
+
+@require_flax
+class BartHeadTests(unittest.TestCase):
```
Comment (patil-suraj) [inline tests/test_modeling_flax_mbart.py:234]:
```suggestion
class MBartHeadTests(unittest.TestCase):
```

File: tests/test_modeling_flax_mbart.py [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+    # @timeout_decorator.timeout(1)  # not working with the decorator so far
+    def test_lm_forward(self):
+        config, input_ids, batch_size = self._get_config_and_data()
+        lm_model = FlaxMBartForConditionalGeneration(config)
+        outputs = lm_model(input_ids=input_ids)
+        expected_shape = (batch_size, input_ids.shape[1], config.vocab_size)
+        self.assertEqual(outputs["logits"].shape, expected_shape)
+
+    def test_lm_uneven_forward(self):
+        config = BartConfig(
```
Comment (patil-suraj) [inline tests/test_modeling_flax_mbart.py:298]:
```suggestion
        config = MBartConfig(
```

File: tests/test_modeling_flax_mbart.py [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+                [26, 1, 60, 48, 22, 13, 2],
+                [21, 5, 62, 28, 14, 76, 2],
+                [45, 98, 37, 86, 59, 48, 2],
+                [70, 70, 50, 9, 28, 0, 2],
+            ],
+            dtype=np.int64,
+        )
+
+        batch_size = input_ids.shape[0]
+        config = BartConfig(
```
Comment (patil-suraj) [inline tests/test_modeling_flax_mbart.py:258]:
```suggestion
        config = MBartConfig(
```

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for adding this @stancld!

It's looking great overall, I left a few comments. Specifically

- The order of layer norm and attention and
- Could we add as much copied from statements as possible?

**Comment by stancld** [comment]
@patil-suraj Thank you a lot for your suggestions. I fixed the order of attention and normalization layers and some other minor bugs. Also added some additional copy statements.

I also changed the `shift_tokens_right` method as this one looks to be different for the MBart models as they don't have a single `decoder_start_token_id` in contrast to other Bart-like models. => This change, however, currently leads to some issues within the `generate` method.

File: src/transformers/generation_flax_utils.py [hunk 10]
```diff
@@ -238,14 +238,14 @@ def generate(
...
-
         if self.config.is_encoder_decoder:
             # add encoder_outputs to model_kwargs
             model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(input_ids, model_kwargs)
             # prepare decoder_input_ids for generation
-            input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+            if decoder_start_token_id is not None:
+                input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+            else:
+                input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4")
```
Comment (patil-suraj) [inline src/transformers/generation_flax_utils.py:248]: `decoder_start_token_id` should never be `None`. For mBART it is the target language id, which could be passed an argument to the `generate`. So let's remove this.

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 11]
```diff
@@ -0,0 +1,1744 @@
...
+            residual = hidden_states
+
+            hidden_states, cross_attn_weights = self.encoder_attn(
+                hidden_states=hidden_states,
+                key_value_states=encoder_hidden_states,
+                attention_mask=encoder_attention_mask,
+            )
+            hidden_states = self.dropout_layer(hidden_states, deterministic=deterministic)
+            hidden_states = residual + hidden_states
+            hidden_states = self.encoder_attn_layer_norm(hidden_states)
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:584]: This should also be applied before `encoder_attn`

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 12]
```diff
@@ -0,0 +1,1744 @@
...
+        >>> from transformers import MBartTokenizer, FlaxMBartForConditionalGeneration
+
+        >>> model = FlaxMBartForConditionalGeneration.from_pretrained('facebook/mbart-large-cc25')
+        >>> tokenizer = MBartTokenizer.from_pretrained('facebook/mbart-large-cc25')
+
+        >>> ARTICLE_TO_SUMMARIZE = "My friends are cool but they eat too many carbs."
+        >>> inputs = tokenizer([ARTICLE_TO_SUMMARIZE], max_length=1024, return_tensors='jax')
+
+        >>> # Generate Summary
+        >>> summary_ids = model.generate(inputs['input_ids']).sequences
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:1529]:
```suggestion
        >>> summary_ids = model.generate(inputs['input_ids'], decoder_start_token_id=tokenizer.lang_code_to_id[tgt_lang]).sequences
```

**Review by patil-suraj — APPROVED** [review]
Looks good to me! Just left a couple of comments that need to be taken care of before merging.

I will run all slow tests and push the checkpoints to the hub before merging.

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 13]
```diff
@@ -0,0 +1,1741 @@
...
+        # and adjust num_embeddings appropriately. Other models don't have this hack
+        self.offset = 2
+        self.embed_positions = nn.Embed(
+            self.config.max_position_embeddings + self.offset,
+            embed_dim,
+            embedding_init=jax.nn.initializers.normal(self.config.init_std, self.dtype),
+            dtype=self.dtype,
+        )
+        self.layers = FlaxMBartEncoderLayerCollection(self.config, self.dtype)
+        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:733]:
```suggestion
        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
        self.layer_norm = nn.LayerNorm(dtype=self.dtype)        
```

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 14]
```diff
@@ -0,0 +1,1741 @@
...
+        )
+
+        if not return_dict:
+            return outputs
+
+        return FlaxBaseModelOutput(
+            last_hidden_state=outputs.last_hidden_state,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:772]:
```suggestion
        last_hidden_states = outputs[0]
        last_hidden_states = self.layer_norm(last_hidden_states)

        if not return_dict:
            return (last_hidden_states,) + outputs[1:]

        return FlaxBaseModelOutput(
            last_hidden_state=last_hidden_states,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 15]
```diff
@@ -0,0 +1,1741 @@
...
+        self.offset = 2
+        self.embed_positions = nn.Embed(
+            self.config.max_position_embeddings + self.offset,
+            embed_dim,
+            embedding_init=jax.nn.initializers.normal(self.config.init_std, self.dtype),
+            dtype=self.dtype,
+        )
+
+        self.layers = FlaxMBartDecoderLayerCollection(self.config, self.dtype)
+        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:808]:
```suggestion
        self.layernorm_embedding = nn.LayerNorm(dtype=self.dtype)
        self.layer_norm = nn.LayerNorm(dtype=self.dtype)
```

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 16]
```diff
@@ -0,0 +1,1741 @@
...
+
+        if not return_dict:
+            return outputs
+
+        return FlaxBaseModelOutputWithPastAndCrossAttentions(
+            last_hidden_state=outputs.last_hidden_state,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            cross_attentions=outputs.cross_attentions,
+        )
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:856]:
```suggestion
        last_hidden_states = outputs[0]
        last_hidden_states = self.layer_norm(last_hidden_states)

        if not return_dict:
            return (last_hidden_states,) + outputs[1:]

        return FlaxBaseModelOutput(
            last_hidden_state=last_hidden_states,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

File: tests/test_modeling_flax_mbart.py [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+        self.assertListEqual(self.expected_text, generated_words)
+
+    def translate_src_text(self, **tokenizer_kwargs):
+        model_inputs = self.tokenizer(self.src_text, **tokenizer_kwargs, return_tensors="np")
+        generated_ids = self.model.generate(
+            model_inputs.input_ids,
+            attention_mask=model_inputs.attention_mask,
+            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
+            num_beams=2,
+        )
```
Comment (patil-suraj) [inline tests/test_modeling_flax_mbart.py:457]:
```suggestion
        ).sequences
```

**Review by patil-suraj — COMMENTED** [review]
The encoder and decoder reqiores an extra layer_norm at the end

File: tests/test_modeling_flax_mbart.py [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+    def _assert_generated_batch_equal_expected(self, **tokenizer_kwargs):
+        generated_words = self.translate_src_text(**tokenizer_kwargs)
+        self.assertListEqual(self.expected_text, generated_words)
+
+    def translate_src_text(self, **tokenizer_kwargs):
+        model_inputs = self.tokenizer(self.src_text, **tokenizer_kwargs, return_tensors="np")
+        generated_ids = self.model.generate(
+            model_inputs.input_ids,
+            attention_mask=model_inputs.attention_mask,
+            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
```
Comment (patil-suraj) [inline tests/test_modeling_flax_mbart.py:455]:
```suggestion
            decoder_start_token_id=self.tokenizer.lang_code_to_id["ro_RO"],
            early_stopping=True,
```

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 19]
```diff
@@ -0,0 +1,1748 @@
...
+        last_hidden_states = outputs[0]
+        last_hidden_states = self.layer_norm(last_hidden_states)
+
+        if not return_dict:
+            return (last_hidden_states,) + outputs[1:]
+
+        return FlaxBaseModelOutput(
+            last_hidden_state=last_hidden_states,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
```
Comment (patil-suraj) [inline src/transformers/models/mbart/modeling_flax_mbart.py:862]:
```suggestion
            attentions=outputs.attentions,
            cross_attentions=outputs.cross_attentions,
```

**Comment by patil-suraj** [comment]
@stancld I pushed a couple of commits to add the `layer_norm` in encoder and decoder. Now, all slow tests are passing.
@patrickvonplaten  could you please take a final look?

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 20]
```diff
@@ -0,0 +1,1747 @@
...
+    assert pad_token_id is not None, "self.model.config.pad_token_id has to be defined."
+
+    # replace possible -100 values in labels by `pad_token_id`
+    prev_output_tokens = jnp.where(prev_output_tokens == -100, pad_token_id, input_ids)
+    index_of_eos = (jnp.where(prev_output_tokens != pad_token_id, 1, 0).sum(axis=-1) - 1).reshape(-1, 1)
+    decoder_start_tokens = jnp.array(
+        [prev_output_tokens[i, eos_idx] for i, eos_idx in enumerate(index_of_eos)]
+    ).squeeze()
+    # for loop basically does jax-compatible version of prev_output_tokens[:, 1:] = prev_output_tokens[:, :-1].clone()
+    for i in range(prev_output_tokens.shape[1], 0, -1):
```
Comment (patrickvonplaten) [inline src/transformers/models/mbart/modeling_flax_mbart.py:231]: nice!

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 21]
```diff
@@ -0,0 +1,1747 @@
...
+            )
+
+    def _split_heads(self, hidden_states):
+        return hidden_states.reshape(hidden_states.shape[:2] + (self.num_heads, self.head_dim))
+
+    def _merge_heads(self, hidden_states):
+        return hidden_states.reshape(hidden_states.shape[:2] + (self.embed_dim,))
+
+    @nn.compact
+    def _concatenate_to_cache(self, key, value, query, attention_mask):
```
Comment (patrickvonplaten) [inline src/transformers/models/mbart/modeling_flax_mbart.py:279]: can we use some `# Copied from` statements here?

File: src/transformers/models/mbart/modeling_flax_mbart.py [hunk 22]
```diff
@@ -0,0 +1,1747 @@
...
+            rngs,
+            input_ids,
+            attention_mask,
+            decoder_input_ids,
+            decoder_attention_mask,
+            position_ids,
+            decoder_position_ids,
+        )["params"]
+
+    def init_cache(self, batch_size, max_length, encoder_outputs):
```
Comment (patrickvonplaten) [inline src/transformers/models/mbart/modeling_flax_mbart.py:978]: Add copied from statement?

**Review by patrickvonplaten — APPROVED** [review]
Very nice!
